openwisp_monitoring/db/backends/elasticsearch/index.py

```diff
@@ -0,0 +1,109 @@
+    tags = Nested(dynamic=True, required=False, multi=True)
+    points = Nested(Point)
+
+    class Index:
+        name = 'metric'
+        settings = {
+            'number_of_shards': 1,
+            'number_of_replicas': 0,
+            'lifecycle.name': 'default',
+            'lifecycle.rollover_alias': 'metric',
+        }
+
+
+def find_metric(client, index, tags, retention_policy=None, add=False):
+    search = Search(using=client, index=index)
+    if tags:
+        tags_dict = dict()
+        for key, value in tags.items():
+            tags_dict[f'tags.{key}'] = value
+        q = Q(
+            'nested',
+            path='tags',
+            query=Q(
+                'bool', must=[Q('match', **{k: str(v)}) for k, v in tags_dict.items()]
+            ),
+        )
+    else:
+        q = Q()
+    try:
+        result = list(search.query(q).execute())[0].meta
+        return result['id'], result['index']
+    except (NotFoundError, AttributeError, IndexError):
+        if add:
+            obj = add_doc(client, index, tags, retention_policy=retention_policy)
+            return obj['_id'], obj['_index']
+        return None
+
+
+def add_doc(client, key, tags, _id=None, retention_policy=None):
+    """
+    Add index to elasticsearch using ``keys``, ``tags`` and ``id`` provided.
+    If no ``id`` is provided a random ``uuid`` would be used.
+    """
+    _id = str(_id or uuid.uuid1())
+    # If index exists, add the doc and return
+    try:
+        index_aliases = client.indices.get_alias(index=key)
+        for k, v in index_aliases.items():
+            if v['aliases'][key]['is_write_index']:
+                break
+        client.create(index=k, id=_id, body={'tags': tags})
+        return {'_id': _id, '_index': k}
+    except NotFoundError:
+        pass
+    # Create a new index if it doesn't exist
+    name = f'{key}-000001'
+    obj = MetricIndex(meta={'id': _id})
+    obj._index = obj._index.clone(name)
+    # Create a new index template if it doesn't exist
+    if not client.indices.exists_template(name=key):
+        obj._index.settings(**{'lifecycle.rollover_alias': key})
+        if retention_policy:
+            obj._index.settings(**{'lifecycle.name': retention_policy})
+        # index pattern is added for Index Lifecycle Management
+        obj._index.as_template(key, f'{key}-*').save(using=client)
+    obj.init(using=client, index=name)
+    obj.meta.index = name
+    obj.tags = tags
+    obj.save(using=client, index=name)
+    client.indices.put_alias(index=name, name=key, body={'is_write_index': True})
+    if retention_policy:
+        client.indices.put_settings(
+            body={'lifecycle.name': retention_policy}, index=name
+        )
+    client.indices.put_settings(body={'lifecycle.rollover_alias': key}, index=name)
+    client.indices.refresh(index=key)
+    return obj.to_dict(include_meta=True)
+
+
+def bulk_indexing():
+    """ Index all existing metrics """
+    from ... import timeseries_db
```

Comment:
why can't this be imported at the top of the file

Reply:
This function is not actually required. I had added it initially to index all existing metrics in one go (right now this will happen whenever there's a metric write). I was not sure whether it is useful anymore or not, will remove it if it's having no use when the PR is ready :smiley:

Reply:
ok :+1: